File: Js_Solutions/priority_queue/priority_queue.js
// Implement a priority queue with min binary heap

class Node {
    constructor(val, priority) {
        this.value = val;
        this.priority = priority;
    }
}

class PriorityQueue {
    constructor() {
        this.values = [];
    }

    enqueue(val, priority) {
        const newNode = new Node(val, priority)
        this.values.push(newNode);
        this.bubbleUp();
        return this.values;
    }

    bubbleUp() {
        let idx = this.values.length - 1;
        const element = this.values[idx];
        while (idx > 0) {
            let parentIdx = Math.floor((idx - 1) / 2);
            let parent = this.values[parentIdx];
            if (element.priority >= parent.priority) {
                break;
            }
            this.values[parentIdx] = element;
            this.values[idx] = parent;
            idx = parentIdx;
        }
    }

    dequeue() {
        const min_val = this.values[0];
        const last = this.values.pop();
        if (this.values.length > 0) {
            this.values[0] = last;
            this.sinkDown();
        }
        return min_val;
    }

    sinkDown() {
        let idx = 0;
        const element = this.values[0];
        const length = this.values.length;

        while (true) {
            let leftChildIdx = 2 * idx + 1;
            let rightChildIdx = 2 * idx + 2;
            let leftChild, rightChild;
            let swap = null;

            if (leftChildIdx < length) {
                leftChild = this.values[leftChildIdx];
                if (leftChild.priority < element.priority) {
                    swap = leftChildIdx;
                }
            }

            if (rightChildIdx < length) {
                rightChild = this.values[rightChildIdx];
                if ((swap === null && rightChild.priority < element.priority) || (swap !== null && rightChild.priority < leftChild.priority)) {
                    swap = rightChildIdx;
                }
            }

            if (swap === null) break;
            this.values[idx] = this.values[swap];
            this.values[swap] = element;
            idx = swap;
        }
    }
}

priority = new PriorityQueue();
priority.enqueue("common cold", 3)
priority.enqueue("wound", 1)
console.log(priority.enqueue("high fever", 2));
console.log(priority.dequeue());
console.log(priority);

File: Python_Solutions/priority_queue/priority_queue.py
# Implement a priority queue with min binary heap

class Node:
    def __init__(self, val, priority):
        self.value = val
        self.priority = priority


class PriorityQueue:
    def __init__(self):
        self.values = []

    def enqueue(self, val, priority):
        newNode = Node(val, priority)
        self.values.append(newNode)
        self.bubbleUp()
        return self.values

    def bubbleUp(self):
        idx = len(self.values) - 1
        element = self.values[idx]
        while idx > 0:
            parentIdx = (idx-1)//2
            parent = self.values[parentIdx]

            if element.priority >= parent.priority:
                break
            self.values[idx] = parent
            self.values[parentIdx] = element
            idx = parentIdx

    def dequeue(self):
        min_val = self.values[0]
        last = self.values.pop()
        if len(self.values) > 0:
            self.values[0] = last
            self.sinkDown()
        return min_val

    def sinkDown(self):
        idx = 0
        element = self.values[idx]
        length = len(self.values)

        while True:
            leftChildIdx = 2*idx + 1
            rightChildIdx = 2*idx + 2
            leftChild = None
            rightChild = None
            swap = None

            if leftChildIdx < length:
                leftChild = self.values[leftChildIdx]
                if leftChild.priority < element.priority:
                    swap = leftChildIdx

            if rightChildIdx < length:
                rightChild = self.values[rightChildIdx]
                if (swap is None and rightChild.priority < element.priority) or (swap is not None and rightChild.priority < leftChild.priority):
                    swap = rightChildIdx

            if swap is None:
                break
            self.values[idx] = self.values[swap]
            self.values[swap] = element
            idx = swap


priority = PriorityQueue()
priority.enqueue("common cold", 3)
priority.enqueue("wound", 1)
print(priority.enqueue("high fever", 2))
print(priority.dequeue())
print(priority.values)
